[Woo POS][Local Catalog] Do not run incremental sync with disabled local catalog

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
# Do not run incremental sync with disabled local catalog + refactor
WOOMOB-1564

💡 Don't merge until the parent branch is `trunk`
<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR:
1. Optimizes the `WooPosPerformLocalCatalogIncrementalSync` use case by adding `WooPosIsLocalCatalogSupported` check before attempting to perform incremental sync operation. 
2. Refactors `WooPosPerformLocalCatalogIncrementalSync` to make it a synchronous, testable function with better error handling and introduces comprehensive unit tests:
- Converted from fire-and-forget coroutine to a suspending function that returns `PosLocalCatalogSyncResult?`
- Replaced `WooPosLocalCatalogM1Enabled` feature flag check with `WooPosIsLocalCatalogSupported` for consolidated site-level support checking
- Replaced `AppCoroutineScope` with `CoroutineDispatchers` for testable coroutine management
3. Covers `WooPosPerformLocalCatalogIncrementalSync` with unit tests

### Test Steps
<!-- Describe how to test your changes. Include only what’s needed for the reviewer to validate the behavior.
For new features, outline the main user flow or goal rather than every tap or click — the reviewer should be able to complete the flow naturally.
If applicable, mention key devices, scenarios, or edge cases to verify. -->
1. Verify incremental sync runs as expected: 
- on payment success
- periodically
- upon POS opening

2. Verify incremental sync doesn't work at all in case local catalog is disabled.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
N/A

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Add local catalog eligibility check to WooPosPerformLocalCatalogIncrementalSync

Refactored WooPosPerformLocalCatalogIncrementalSync to improve testability

samiuelson committed:
Add tests

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/localcatalog/WooPosPerformLocalCatalogIncrementalSync.kt:
```diff
@@ -1,55 +1,64 @@
 ) {
-    fun execute(reason: WooPosIncrementalSyncReason) {
+    @Suppress("ReturnCount")
+    suspend fun execute(reason: WooPosIncrementalSyncReason): PosLocalCatalogSyncResult? {
```
❓ Seems like we are returning a result now, but we are not using it anywhere 🤔. 

Moreover, it seems that for example in [this place](https://github.com/woocommerce/woocommerce-android/blob/0c1b2e085cb0deb00690f5be3f8d2bc9eedc67c6/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt#L312) it'll block the VM from collecting any more UI Events until the sync completes.


Is this an intentional change so the logic can be more easily tested or what is the reasoning behind it?

samiuelson replied:
Good catch 😞 I reverted the approach to run in app scope, allowing "fire and forget" incremental sync type. Fixed in: 79bfd63cbac3bee6595b8fad48549658fed9065c

malinajirka reviewed (commented):
Thanks @samiuelson! I've not completed a testing round yet, I'd like to wrap up the below discussion first. Thanks!

samiuelson committed:
Revert runing incremental sync in AppCoroutineScope

Moves the local catalog incremental sync operation off the main thread by launching it within the `AppCoroutineScope`.

This change refactors `WooPosPerformLocalCatalogIncrementalSync` to no longer be a suspending function and removes its direct return value. The operation is now fire-and-forget from the caller's perspective, with logging handling the sync result.

Additionally, tests are updated to reflect this asynchronous behavior, verifying interactions and using `advanceUntilIdle` instead of checking return values.

samiuelson committed:
Refactor WooPosPerformLocalCatalogIncrementalSync to simplify coroutine usage

Removes the explicit `CoroutineDispatchers` dependency and the `withContext(dispatchers.io)` wrapper from `WooPosPerformLocalCatalogIncrementalSync`, as the underlying `localCatalogSyncRepository` already manages its own thread switching.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/localcatalog/WooPosPerformLocalCatalogIncrementalSync.kt:
```diff
@@ -3,53 +3,65 @@ package com.woocommerce.android.ui.woopos.localcatalog
-                    is Failure -> {
-                        wooPosLogWrapper.e("Sync $reasonDescription failed: ${syncResult.error}")
-                    }
+            val site = selectedSite.getOrNull()
+            if (site == null) {
+                wooPosLogWrapper.d("Skipping sync $reasonDescription: No site selected")
+                return@launch
+            }
+
+            if (!isLocalCatalogSupported(site.localId())) {
```
Interesting, but I think we actually shouldn't wrap any calls with `withContext(IO)` inside `WooPosPerformLocalCatalogIncrementalSync`. At least this is a convention we have in the project. 

I removed the `withContext(IO)` from `WooPosPerformLocalCatalogIncrementalSync::performSync` because `WooPosLocalCatalogSyncRepository::syncLocalCatalogIncremental` uses `withContext(IO)` internally: [8ab3a6c]. 

I believe The same should be done for `WooPosIsLocalCatalogSupported` it (or its dependencies) should control dispatchers on which they access DB or remote APIs.

samiuelson replied:
This PR updates WooPosIsLocalCatalogSupported to use IO dispatcher:  https://github.com/woocommerce/woocommerce-android/pull/14904

malinajirka reviewed (approved):
LGTM, thanks!
